### ISSUE TITLE
Unable to connect

### ISSUE THREAD

**Issue by modul** [dagargo/microdude#10]
First of all, thank you for doing this! 

Unfortunately I'm unable to connect to my Microbrute:

```
~ $ microdude -v
DEBUG:microdude.connector:Initializing...
DEBUG:microdude.utils:Reading config file...
DEBUG:microdude.utils:Config file read.
DEBUG:microdude.connector:Connecting to ...
DEBUG:microdude.connector:Mido backend: <backend mido.backends.rtmidi (loaded)>
DEBUG:microdude.connector:Handshaking...
DEBUG:microdude.connector:Sending message 0x7e, 0x7f, 0x6, 0x1...
DEBUG:microdude.connector:Receiving message 0x7e, 0x7f, 0x6, 0x1...
DEBUG:microdude.connector:Bad handshake. Disconnecting...
DEBUG:microdude.connector:Disconnecting...
INFO:microdude.editor:Not connected
```

```
~ $ lsusb
Bus 002 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 001 Device 003: ID 8087:0a2b Intel Corp. 
Bus 001 Device 002: ID 0bda:57e1 Realtek Semiconductor Corp. Lenovo EasyCamera
Bus 001 Device 004: ID 1c75:0402 Arturia MicroBrute
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
```

Am I missing something?

**Comment by dagargo** [comment]
It's nice you find MicroDude useful.

Your problem is in the handshake. For some reason, it's trying to connect to a "" named device. Besides, it looks like the device is echoing the input.

Here you have my start up messages. As you see my device name is "MicroBrute:MicroBrute MIDI 1 20:0".
```
$ microdude -v
DEBUG:microdude.connector:Initializing...
DEBUG:microdude.utils:Reading config file...
DEBUG:microdude.utils:Config file read.
DEBUG:microdude.connector:Connecting to MicroBrute:MicroBrute MIDI 1 20:0...
DEBUG:microdude.connector:Mido backend: <backend mido.backends.rtmidi (loaded)>
DEBUG:microdude.connector:Handshaking...
DEBUG:microdude.connector:Sending message 0x7e, 0x7f, 0x6, 0x1...
DEBUG:microdude.connector:Receiving message 0x7e, 0x1, 0x6, 0x2, 0x0, 0x20, 0x6b, 0x4, 0x0, 0x2, 0x1, 0x1, 0x0, 0x4, 0x0...
DEBUG:microdude.connector:Handshake ok. Version 1.0.4.0.
[...]
```

lsusb seems ok. Can you post the output of your `aconnect -lio`? This is mine.
```
$ aconnect -lio
client 0: 'System' [type=kernel]
    0 'Timer           '
    1 'Announce        '
client 14: 'Midi Through' [type=kernel]
    0 'Midi Through Port-0'
client 20: 'MicroBrute' [type=kernel,card=1]
    0 'MicroBrute MIDI 1'
    1 'MicroBrute MIDI 2'
```

Could you post this also?
```
$ cat ~/.microdude/config 
{"device": "MicroBrute:MicroBrute MIDI 1 20:0", "persistent": false}
```

What devices do you see at the preferences window? I've been able to replicate your output minus the device name by selecting the "Midi Through:Midi Through Port-0 14:0" device.

**Comment by modul** [comment]
OK so first of all, I didn't realize that the preferences tab window was enabled (sorry for that), since it doesn't look that way:

![2020-01-16-225957_652x472_scrot](https://user-images.githubusercontent.com/700246/72566421-11ac6180-38b4-11ea-80fe-4eedad2fb719.png)

I was able to see the Microbrute in the dropdown menu (2 channels + through port) as well as with `aconnect -lio`:

```
client 0: 'System' [type=Kernel]
    0 'Timer           '
    1 'Announce        '
client 14: 'Midi Through' [type=Kernel]
    0 'Midi Through Port-0'
client 20: 'MicroBrute' [type=Kernel,card=1]
    0 'MicroBrute MIDI 1'
    1 'MicroBrute MIDI 2'
```

The first port seems to be working but than I still get a python exception:

```
~ $ microdude -v
DEBUG:microdude.connector:Initializing...
DEBUG:microdude.utils:Reading config file...
DEBUG:microdude.utils:Config file read.
DEBUG:microdude.connector:Connecting to MicroBrute:MicroBrute MIDI 2 20:1...
DEBUG:microdude.connector:Mido backend: <backend mido.backends.rtmidi (loaded)>
DEBUG:microdude.connector:Handshaking...
DEBUG:microdude.connector:Sending message 0x7e, 0x7f, 0x6, 0x1...
DEBUG:microdude.connector:Receiving message 0x7e, 0x1, 0x6, 0x2, 0x0, 0x20, 0x6b, 0x4, 0x0, 0x2, 0x2, 0x1, 0x0, 0x0, 0x0...
DEBUG:microdude.connector:Bad handshake. Disconnecting...
DEBUG:microdude.connector:Disconnecting...
INFO:microdude.editor:Not connected
DEBUG:microdude.editor:Adding port MicroBrute:MicroBrute MIDI 1 20:0...
DEBUG:microdude.editor:Adding port MicroBrute:MicroBrute MIDI 2 20:1...
DEBUG:microdude.editor:Port MicroBrute:MicroBrute MIDI 2 20:1 is active
DEBUG:microdude.editor:Adding port Midi Through:Midi Through Port-0 14:0...
DEBUG:microdude.editor:Configuration: {'device': 'MicroBrute:MicroBrute MIDI 1 20:0', 'persistent': True}
DEBUG:microdude.utils:Writing config file...
DEBUG:microdude.utils:Config file written.
DEBUG:microdude.connector:Connecting to MicroBrute:MicroBrute MIDI 1 20:0...
DEBUG:microdude.connector:Mido backend: <backend mido.backends.rtmidi (loaded)>
DEBUG:microdude.connector:Handshaking...
DEBUG:microdude.connector:Sending message 0x7e, 0x7f, 0x6, 0x1...
DEBUG:microdude.connector:Receiving message 0x7e, 0x1, 0x6, 0x2, 0x0, 0x20, 0x6b, 0x4, 0x0, 0x2, 0x1, 0x1, 0x0, 0x0, 0x0...
DEBUG:microdude.connector:Handshake ok. Version 1.0.0.0.
DEBUG:microdude.connector:Sending message 0x0, 0x20, 0x6b, 0x5, 0x1, 0x0, 0x0, 0x6...
DEBUG:microdude.connector:Receiving message 0x0, 0x20, 0x6b, 0x5, 0x1, 0x0, 0x1, 0x5, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0...
INFO:microdude.editor:Connected (firmware version 1.0.0.0)
DEBUG:microdude.editor:Loading status...
DEBUG:microdude.connector:Sending message 0x0, 0x20, 0x6b, 0x5, 0x1, 0x1, 0x0, 0x6...
DEBUG:microdude.connector:Receiving message 0x0, 0x20, 0x6b, 0x5, 0x1, 0x1, 0x1, 0x5, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0...
DEBUG:microdude.connector:Sending message 0x0, 0x20, 0x6b, 0x5, 0x1, 0x2, 0x0, 0x8...
DEBUG:microdude.connector:Receiving message 0x0, 0x20, 0x6b, 0x5, 0x1, 0x2, 0x1, 0x7, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0...
DEBUG:microdude.connector:Sending message 0x0, 0x20, 0x6b, 0x5, 0x1, 0x3, 0x0, 0x35...
DEBUG:microdude.connector:Receiving message 0x0, 0x20, 0x6b, 0x5, 0x1, 0x3, 0x1, 0x34, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1...
DEBUG:microdude.connector:Sending message 0x0, 0x20, 0x6b, 0x5, 0x1, 0x4, 0x0, 0x10...
DEBUG:microdude.connector:Receiving message 0x0, 0x20, 0x6b, 0x5, 0x1, 0x4, 0x1, 0xf, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0...
DEBUG:microdude.connector:Sending message 0x0, 0x20, 0x6b, 0x5, 0x1, 0x5, 0x0, 0x2f...
DEBUG:microdude.connector:Receiving message 0x0, 0x20, 0x6b, 0x5, 0x1, 0x5, 0x1, 0x2e, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1...
DEBUG:microdude.connector:Sending message 0x0, 0x20, 0x6b, 0x5, 0x1, 0x6, 0x0, 0xc...
DEBUG:microdude.connector:Receiving message 0x0, 0x20, 0x6b, 0x5, 0x1, 0x6, 0x1, 0xb, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0...
DEBUG:microdude.connector:Sending message 0x0, 0x20, 0x6b, 0x5, 0x1, 0x7, 0x0, 0xe...
DEBUG:microdude.connector:Receiving message 0x0, 0x20, 0x6b, 0x5, 0x1, 0x7, 0x1, 0xd, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0...
DEBUG:microdude.connector:Sending message 0x0, 0x20, 0x6b, 0x5, 0x1, 0x8, 0x0, 0x12...
DEBUG:microdude.connector:Receiving message 0x0, 0x20, 0x6b, 0x5, 0x1, 0x8, 0x1, 0x11, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0...
DEBUG:microdude.connector:Sending message 0x0, 0x20, 0x6b, 0x5, 0x1, 0x9, 0x0, 0x33...
DEBUG:microdude.connector:Receiving message 0x0, 0x20, 0x6b, 0x5, 0x1, 0x9, 0x1, 0x32, 0x2, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0...
DEBUG:microdude.connector:Sending message 0x0, 0x20, 0x6b, 0x5, 0x1, 0xa, 0x0, 0x2d...
DEBUG:microdude.connector:Disconnecting...
Traceback (most recent call last):
  File "/home/mo/.local/lib/python3.8/site-packages/MicroDude-1.5-py3.8.egg/microdude/editor.py", line 91, in <lambda>
  File "/home/mo/.local/lib/python3.8/site-packages/MicroDude-1.5-py3.8.egg/microdude/editor.py", line 115, in save
  File "/home/mo/.local/lib/python3.8/site-packages/MicroDude-1.5-py3.8.egg/microdude/editor.py", line 258, in ui_reconnect
  File "/home/mo/.local/lib/python3.8/site-packages/MicroDude-1.5-py3.8.egg/microdude/editor.py", line 285, in set_ui
  File "/home/mo/.local/lib/python3.8/site-packages/MicroDude-1.5-py3.8.egg/microdude/connector.py", line 266, in get_parameter
  File "/home/mo/.local/lib/python3.8/site-packages/MicroDude-1.5-py3.8.egg/microdude/connector.py", line 341, in rx_message
microdude.connector.ConnectorError: Connection error
```

The second port gives me a bad handshake:

```
DEBUG:microdude.editor:Configuration: {'device': 'MicroBrute:MicroBrute MIDI 2 20:1', 'persistent': True}
DEBUG:microdude.utils:Writing config file...
DEBUG:microdude.utils:Config file written.
DEBUG:microdude.connector:Connecting to MicroBrute:MicroBrute MIDI 2 20:1...
DEBUG:microdude.connector:Mido backend: <backend mido.backends.rtmidi (loaded)>
DEBUG:microdude.connector:Handshaking...
DEBUG:microdude.connector:Sending message 0x7e, 0x7f, 0x6, 0x1...
DEBUG:microdude.connector:Receiving message 0x7e, 0x1, 0x6, 0x2, 0x0, 0x20, 0x6b, 0x4, 0x0, 0x2, 0x2, 0x1, 0x0, 0x0, 0x0...
DEBUG:microdude.connector:Bad handshake. Disconnecting...
DEBUG:microdude.connector:Disconnecting...
INFO:microdude.editor:Not connected
```

And the through port is echoing:

```
DEBUG:microdude.editor:Configuration: {'device': 'Midi Through:Midi Through Port-0 14:0', 'persistent': True}
DEBUG:microdude.utils:Writing config file...
DEBUG:microdude.utils:Config file written.
DEBUG:microdude.connector:Connecting to Midi Through:Midi Through Port-0 14:0...
DEBUG:microdude.connector:Mido backend: <backend mido.backends.rtmidi (loaded)>
DEBUG:microdude.connector:Handshaking...
DEBUG:microdude.connector:Sending message 0x7e, 0x7f, 0x6, 0x1...
DEBUG:microdude.connector:Receiving message 0x7e, 0x7f, 0x6, 0x1...
DEBUG:microdude.connector:Bad handshake. Disconnecting...
DEBUG:microdude.connector:Disconnecting...
INFO:microdude.editor:Not connected
```

**Comment by dagargo** [comment]
Having black icons and black themes makes the UI quite unusable. To me, it looks like the first, fourth and sixth icons are even darker, which for a dark icon theme means they are active. Try to change the icon theme to improve this.

Focusing on the handshaking problem, it seems you fixed it by choosing the "MicroBrute:MicroBrute MIDI 1 20:0" device. This is the one I'm using and the only one that works. I don't know why there are 2 interfaces but this is the one you need.

But it's still not working.
Looks like your Microbrute is using firmware 1.0.0.0, which I've never used and is quite old. Would it be possible for you to upgrade it to a later version? You'll need an official Arturia tool to do so as MicroDude can't be used to upgrade the firmware.

**Comment by modul** [comment]
It's funny, I never had this problem on other UIs, but anyway.

I'll update the firmware some way. Looks like this will fix everything.

Thanks for your help. And again, thanks for this tool. :)

**Comment by dagargo** [comment]
I've tried to change the icon themes but didn't found one that matches your screenshot.
Could you please provide some information about your setup (distribution, desktop environment...) in order to try to replicate the icon issue? I use Debian stable with Gnome and runs fine.

Anyway, we might forget about this. I've been thinking of removing the icon bar and putting all the functionality in the application menu. May I ask what do you think about this?

**Comment by modul** [comment]
I'm using KDE/Plasma 5 on Arch with the standard dark theme called Breeze; icon theme is also Breeze, even for Gtk 3. So it shouldn't be anything fancy. 

I'm sure moving all functionality into the menu bar or a single menu would work just as well. I'd prefer it, but don't loose the icons just on my account.  :)

**Comment by dagargo** [comment]
In case you're still interested, I tried MicroDude on Kubuntu and it worked nicely both on the default clear mode and also in dark mode. The only thing I had to do was setting the GTK3 and GTK2 themes and icons to breeze dark in an application called "GNOME/GTK Application Style".
There is no much difference between sensitive and unsensitive icons but at least the contrast make them quite usable.

I will definitely move all the menu icons to the application in the near future.
